src/lua/compat.lua

```diff
@@ -0,0 +1,307 @@
+                    result = result .. '    '
+                end
+                result = result .. ' -- ' .. comment
+            end
+        end
+        if not result:find('\n') then
+            return result .. '})'
+        else
+            -- Need double '\n' due to gh-3012.
+            return result .. '\n})\n\n'
+        end
+    end,
+    add_option = function(policy)
+        if type(policy) ~= 'table' then
+            error("usage: compat.add_option({name = '...', default = 'new'/'old'," ..
+                  " brief = '...', action = func, run_action = true/false})")
+        end
+        local name = policy.name
+        verify_option(name, policy)
+
+        local current = policy.default == 'new'
+        if not options[name] then
+            options[name] = {
+                current = current,
+                selected = false
+            }
+            table.insert(options_order, name)
+        elseif not options[name].selected then
+            options[name].current = current
+        end
+
+        local option = options[name]
+        option.brief    = policy.brief
+        option.default  = policy.default
+        option.obsolete = policy.obsolete
+        option.action   = policy.action
+
+        if not option.obsolete and policy.run_action then
+            option.action(options[name].current)
+        end
+    end,
+    preload = function()
+        for key, option in pairs(options) do
+            verify_option(key, option)
+            table.insert(options_order, key)
+            option.current = option.default == 'new'
+            option.selected = false
+        end
+    end,
+    postload = function()
+        for _key, option in pairs(options) do
+            if not option.obsolete and option.run_action then
+                option.action(option.current)
+            end
+        end
+    end,
+    help = function()
+        return help
+    end
+}, {
+    __call = function(_, list)
+        if type(list) ~= 'table' then
+            error("usage: compat({option1 = 'new'/'old'/'default'})")
+        end
+        for key, val in pairs(list) do
+            set_option(key, val)
+        end
+    end,
+    __newindex = function(_, key, val)
+        set_option(key, val)
+    end,
+    __index = function(_, key)
+        if not options[key] then
+            error(('Invalid option %s'):format(key))
+        end
+        return serialize_option(key)
+    end,
+    __serialize = serialize_compat,
+    __tostring  = function()
+        return require('yaml').encode(serialize_compat())
+    end,
+    __autocomplete = function()
+        local res = { }
+        for key, option in pairs(options) do
+            if not option.obsolete then
+                res[key] = true
+            end
+        end
+        return res
+    end
+})
+
+-- This functions are to be run only once, thus they are removed not to confuse users.
+compat.preload()
+compat.preload = nil
+compat.postload()
+compat.postload = nil
```

Comment:
It had been left from the time when we didn't know when it is safe (in terms of loaded modules) to run action. I didn't want to remove it in case we rethink it once again before or after the merge.

Reply:
Quorum decision (with @igormunkin): let's move them into local function and properly comment why we split them.

Reply:
Done.

Reply:
> BTW, is not it the same as calling compat.add_option(<...>) for all built-in options?

I would like to see an answer.

Reply:
It is somewhat like, but table format is a bit different, no hot reload logic, no `action` call and no `run_action` stripping.

Reply:
It seems, we don't get each other at first. I asked details from Gleb in a chat and the resume it the following.

It is possible to add built-in options using `add_option()`. AFAIU, it'll perform some unneeded steps, but that's harmless. It'll also contradict preload/postload splitting, but, TBH, I don't understand why they're splitted.

We agreed that it is a matter of taste. I think that built-in and external options should use the same `add_option()` API and adding built-in options using a separate mechanic is error-prone. But that's my personal opinion.

I'll leave this discussion open, maybe @igormunkin wants to participate. But if he don't want, let's stick with Gleb's decision. Let's consider this discussion as not a blocker.